src/main/java/software/amazon/event/kafkaconnector/offloading/S3EventBridgeEventDetailValueOffloading.java

```diff
@@ -0,0 +1,223 @@
+  }
+
+  public static void validateJsonPath(String jsonPathExp) {
+    transformToPutEventsRequestEntry(jsonPathExp);
+  }
+
+  @Override
+  public EventBridgeMappingResult apply(
+      final List<MappedSinkRecord<PutEventsRequestEntry>> putEventsRequestEntries) {
+
+    var result =
+        putEventsRequestEntries.stream()
+            .map(this::apply)
+            .collect(partitioningBy(EventBridgeResult::isSuccess));
+
+    var success = result.get(true).stream().map(EventBridgeResult::success).collect(toList());
+    var errors = result.get(false).stream().map(EventBridgeResult::failure).collect(toList());
+
+    return new EventBridgeMappingResult(success, errors);
+  }
+
+  private EventBridgeResult<PutEventsRequestEntry> apply(
+      final MappedSinkRecord<PutEventsRequestEntry> item) {
+
+    var sinkRecord = item.getSinkRecord();
+    var putEventsRequestEntry = item.getValue();
+
+    try {
+      var documentContext = JsonPath.parse(putEventsRequestEntry.detail(), jsonPathConfiguration);
+
+      var value = documentContext.read(jsonPathRemove);
+      if (value != null) {
+        var payload =
+            value instanceof Map || value instanceof List
+                ? JsonPath.parse(value).jsonString()
+                : value.toString();
+        documentContext
+            .delete(jsonPathRemove)
+            .put(jsonPathAdd, "dataref", s3ArnOf(putS3Object(payload)));
+      }
+      documentContext.put(jsonPathAdd, "datarefJsonPath", jsonPathExp);
+
+      return success(
+          sinkRecord, putEventsRequestEntry.copy(it -> it.detail(documentContext.jsonString())));
+
+    } catch (final ThrowingFunctionApplyException e) {
+      var unwrapped = e.getCause();
+      if (unwrapped instanceof ExecutionException
+          || unwrapped instanceof InterruptedException
+          || unwrapped instanceof TimeoutException) {
+
+        var cause = unwrapped.getCause();
+        if (cause instanceof S3Exception && ((S3Exception) cause).statusCode() < 500) {
+          return failure(sinkRecord, panic(e));
+        }
+        return failure(sinkRecord, retry(e));
+      }
+      return failure(sinkRecord, panic(unwrapped));
+    }
+  }
+
+  private String s3ArnOf(UUID uuid) {
+    return format("arn:aws:s3:::%s/%s", bucketName, uuid);
+  }
+
+  private UUID putS3Object(final String payload) {
+    var durationHolder = new Duration[] {null};
+    var s3ObjectId =
+        s3ObjectKeyCache.computeIfAbsent(
+            sha512Of(payload),
+            ThrowingFunction.wrap(
+                (sha512) -> {
+                  var id = idGenerator.get();
+                  var request =
+                      PutObjectRequest.builder().bucket(bucketName).key(id.toString()).build();
+
+                  var body = fromString(payload, UTF_8);
+
+                  var start = OffsetDateTime.now(UTC);
+                  client.putObject(request, body).get(SDK_TIMEOUT, MILLISECONDS);
```

Comment:
nit: can I get a TRACE log before indicating the call is being made, something like "uploading object to s3 with key {}" with placeholders for the key?